### ISSUE TITLE
SecretService client authorisation for short-lived processes

### ISSUE THREAD

**Issue by nbarrientos** [keepassxreboot/keepassxc#7571]
Hi,

If I understand correctly, since 2.7.0, apps requesting secrets via SecretService have to be authorised to do so, either every time a secret request is issued (`Allow Selected`) or permanently (`Allow All and Future`). I understand the value of this feature which strengthens KeepassXC's security. 

However, in my case [isync](https://isync.sourceforge.io/) uses `secret-tool` to retrieve the password of my IMAP account, with something along the lines of:

```
PassCmd "/usr/bin/secret-tool lookup host mail.example.org account joe@example.org"
```

The issue now is that every run of the mail sync process creates a new `secret-tool` process so KeypassXC asks interactively for authorisation to deliver the secret. I imagine that the SecretService implementation is designed to save the PID or other fingerprint of a running process and whitelist it and does not have short-lived processes in mind (just a guess, sorry, I haven't looked at the implementation details).

![image](https://user-images.githubusercontent.com/131893/159458370-2c4a6289-6bca-4c49-b007-0b5eba42534d.png)

As you can imagine, this is a bit cumbersome. I've set for the time being `ConfirmAccessItem=false`, however maybe it'd be nice to have more fine grained authorisation rules like: "all requests coming from a process with binary `/usr/bin/secret-tool` and parent `/usr/bin/mbsync` are authorised to retrieve secret A". I'm rather happy with `ConfirmAccessItem=false` as, if I recall correctly, it restores the behaviour of the previous version of KeepassXC however, does my proposal make any sense to you?

Thanks.

**Comment by droidmonkey** [comment]
Hmmm, I could imagine remembering the process stack instead of, or in addition to, pid. This would let you remember that a specific process was using secret-tool to request credentials. Perhaps could be a special case for secret-tool only.

**Comment by nbarrientos** [comment]
Yup, in my case it might be tricky because it's `systemd` spawning `mbsync` via a service unit and a timer so the only "resident" process to remember/authorise in the tree would be `systemd --user` which is a bit too generic. Ideally, for me having a way to define a rule to allow any requests from `/usr/bin/secret-tool` (any PID) invoked by `/usr/bin/mbsync` (any PID) would be perfect.

Indeed as you're suggesting an alternative would be to handle ``/usr/bin/secret-tool`` as a special case and add another config option to always authorise new `/usr/bin/secret-tool`  processes, regardless of the parent or process stack.

**Comment by CrfzdPQM6** [comment]
I had the same problem using secret-tool to provide login to a caldav server. Another oddity is that every time an access request is triggered, keepass *first* insists on unlocking all databases that are in tabs within the KeePassXC application, and *only then* goes to the one (that was in scope at first) that has the Secret Service access available. I only encountered this before when another tab was active, different to the one that had the Secret Service integration.

**Comment by droidmonkey** [comment]
It is 'dangerous' to trust secret-tool implicitly because any adversary or malicious program would be using it to harvest credentials. So this is a hard problem to solve correctly.

**Comment by jonasc** [comment]
> It is 'dangerous' to trust secret-tool implicitly because any adversary or malicious program would be using it to harvest credentials. So this is a hard problem to solve correctly.

Could a slightly more secure way be to have dedicated scripts which only access a specific entry? Like `/usr/bin/local/get-mail-pwd` as

~~~.bash
#!/usr/bin/bash
exec secret-tool lookup Uuid <my-uuid-here>
~~~

This script could, for additional security, check whether it is called by the right script (e.g. by examining `/proc/$PPID/exe` and `/proc/$PPID/cmdline`).

~~~.bash
#!/usr/bin/bash
[[ $(readlink "/proc/$PPID/exe") == /usr/bin/mbsync ]] || exit 1
exec secret-tool lookup Uuid <my-uuid-here>
~~~

Obviously this script should not be writable as that would defeat the security improvement.

**Comment by Aetf** [comment]
Isn't this #6458?

**Comment by droidmonkey** [comment]
Somewhat, perhaps a special case for explicit approvals of secret-tool parent process.

**Comment by Aetf** [comment]
To special handle secret-tool we have to figure out a way to identify secret-tool. Obviously just checking for
`/usr/bin/secret-tool` isn't reliable due to reasons in #6458.

**Comment by droidmonkey** [comment]
Well you leave it up to the user, just detect the process name (not path):

> Warning: It looks like a process is using secret-tool, be sure you trust the parent process calling it before allowing.

**Comment by Aetf** [comment]
The process name can be even easier to change, e.g. just rename a script to `secret-tool`. So I think client
fingerprinting is still necessary if we ever want to special case secret-tool. And that seems to be doable now. See my other comment there.

We can then decide whether we want to go with the full client fingerprinting and rule matching,
or just special case secret-tool, and check the parent process instead, and do not save persistent decisions like what
we have now. (This seems to be not enough for @nbarrientos, though. As the parent process is also transient)

**Comment by CrfzdPQM6** [comment]
Is it only `secret-tool` that is relevant here? Though, like @nbarrientos , I first came across this issue with `secret-tool`, I've since seen similar behaviour with `pinentry-gnome3`, which handles my GPG key passphrase entry from keepassxc.

**Comment by nbarrientos** [comment]
What about whitelisting the hash of a given binary in the process tree executed by a given user? In my case I'd accept secret retrievals performed by invocations of `/usr/bin/mbsync` (hash of the binary to be saved when whitelisted for the first time) executed by my username. Then if `/usr/bin/mbsync` calls `/usr/bin/secret-tool` or talks via D-Bus directly would be irrelevant.  KeepassXC would have to walk back the process tree looking for a binary hash in the whitelist, basically.

In other words, "allow any process of this trusted binary with hash `h` called by euid `u` (or any child process it creates) to retrieve secret `s`". That would cover my case, I think.

Binary upgrades due to a software updates would have to be re-whitelisted but that's acceptable and much less cumbersome than having to approve every request.

(Thanks for looking at this, btw)

**Comment by ioogithub** [comment]
I am confused with this issue. Is this the same issue as my bug report that was closed here: https://github.com/keepassxreboot/keepassxc/issues/7623?

I have been using the secret service function reliably for over a year. I got this new version of keepass during a dist-upgrade and now I am prompted hundreds of times a day for this Access Request.

I click the "Allow All & Future" button and the blue "Remember" box is checked however the very next email I open Keepass prompts me again for the Access Request. How is this not a bug? Keepass is not remembering the "Remember" checkbox, is simple disregards it. 

This worked perfectly in the previous versions. 2.7.0 totally breaks my daily computing experience.

**Comment by Aetf** [comment]
@ioogithub due to various limitations (mostly discussed in #6458), KPXC only remembers the decision for as long as both the client and the KPXC processes are running.

This works fine for long-running clients, but unfortunately, it turns out many clients simply use a separate short-lived process (like`secret-tool`) to request the password, thus this issue.

You can simply disable "Confirm when passwords are retrieved by clients" to revert to 2.6.x behavior for now.

**Comment by izbushka** [comment]
> You can simply disable "Confirm when passwords are retrieved by clients" to revert to 2.6.x behavior for now.

The problem is this option change nothing since 2.7.2. It still asks for access every time regardless of this setting.

**Comment by mormegil-cz** [comment]
I commented on a wrong bug (closed duplicate), so copying my comment it from there:

I believe the problem with the program ignoring the _Confirm when passwords are retrieved by clients_ setting is in `UnlockPrompt::unlockItems` where the code tests whether `client->itemKnown(uuid)` and then if `!client->itemAuthorized(uuid)`. However: while `DBusClient::itemAuthorized` takes into account the `FdoSecrets::settings()->confirmAccessItem()` setting, `DBusClient::itemKnown` does not.

Which means the code in `unlockItems` ignores the setting and always needs to ask for the confirmation.

I am not sure which of the two places is the correct one to add the check: Either `DBusClient::itemKnown` should return `true` always when `FdoSecrets::settings()->confirmAccessItem()` is unset, or `UnlockPrompt::unlockItems` should check for the setting as well. I tested the latter variant and this seem to work:

```diff
diff --git a/src/fdosecrets/objects/Prompt.cpp b/src/fdosecrets/objects/Prompt.cpp
index bd01de89..e89cd499 100644
--- a/src/fdosecrets/objects/Prompt.cpp
+++ b/src/fdosecrets/objects/Prompt.cpp
@@ -23,6 +23,7 @@
 #include "fdosecrets/objects/Session.h"
 #include "fdosecrets/widgets/AccessControlDialog.h"
 
+#include "FdoSecretsSettings.h"
 #include "core/Entry.h"
 #include "gui/MessageBox.h"
 
@@ -298,7 +299,7 @@ namespace FdoSecrets
                 }
                 auto entry = item->backend();
                 auto uuid = entry->uuid();
-                if (client->itemKnown(uuid)) {
+                if (client->itemKnown(uuid) || !FdoSecrets::settings()->confirmAccessItem()) {
                     if (!client->itemAuthorized(uuid)) {
                         m_numRejected += 1;
```

**Comment by siddhpant** [comment]
The easiest way would be to have keepassxc have its own `secret-tool` implementation which uses the existing running process, and symlink that to `/usr/bin/secret-tool`. It can be as easy as having a script which calls it with appropriate arguments. For example: `keepassxc secret-tool $@`

**Comment by siddhpant** [comment]
> In other words, "allow any process of this trusted binary with hash h called by euid u (or any child process it creates) to retrieve secret s". That would cover my case, I think.

@nbarrientos IIUC, the issue here is malicious secret-tool, so allowing all child-processes doesn't address it. Though I agree whitelisting parent process is a far cleaner way. If someone is able to replace secret-tool they are anyways inside the system.

**Comment by KitchM** [comment]
A lot of gobbledegoop to say that it doesn't work IMHO.  When I start an application on my computer, I expect it to know that I logged in on my desktop on my computer.  When I start an application, I expect it to be recognized because I started it.

Just what is the bottom line?  Why is the secret service interrupting my use of my hardware and software?

**Comment by droidmonkey** [comment]
Pretty clearly described here: https://github.com/keepassxreboot/keepassxc/issues/7571#issuecomment-1076921523

**Comment by KitchM** [comment]
No, that is not correct.  Here's why.

The reader must first go to your reference at "FdoSecrets: client fingerprinting #6458".  Then they must look up FdoSecrets to understand what that means.  That goes to https://github.com/keepassxreboot/keepassxc/issues/6458, and the reader must then understand FDO, so eventually one gets to FIDO and the FIDO Alliance.  Then one reads the document entitled "FDO Security Requirements" to learn the following:
2.1. Introduction
This document assesses the FIDO Device Onboard (FDO) protocol and specifies the security requirements that must be fulfilled in order to attain the stated security goals of this protocol.
2.1.1. Definition
An automatic onboarding protocol for IoT devices permits late binding of device credentials, so that one manufactured device may onboard, without modification, to many different IOT platforms.

This leads to learning about "onboarding", etc., etc......  That rabbit hole runs deep.  And I still do not understand how any of this has to do with the very real problem of sign-in one time and be done with it.

This is a perfect example of why I had to make the documentation on how to install the program correctly in the first place.  There are way too many unclear and undefined parts of the whole thing.

**Comment by mormegil-cz** [comment]
> This leads to learning about "onboarding", etc., etc...... That rabbit hole runs deep. And I still do not understand how any of this has to do with the very real problem of sign-in one time and be done with it.

Me neither. Why do you think this has anything to do with FIDO? I understood your request in that you do not want to confirm every time whether a program should be allowed access to secrets. Was that the question?

If yes, then uncheck the “Confirm when passwords are retrieved by clients” mentioned in the linked comment, and that’s it, isn’t it?

If not, could you elaborate the question a bit?

**Comment by michaelk83** [comment]
FdoSecrets is just another name for Secret Service. Fdo here is short for FreeDesktop.org, which is where the Secret Service API was defined. It has nothing to do with FIDO.

You don't need to do a deep read. All of the information is in the [linked comment](https://github.com/keepassxreboot/keepassxc/issues/7571#issuecomment-1076921523) itself:

> due to various limitations (_technical details not important for end user_), **KPXC (KeePassXC) only remembers the decision for as long as both the client and the KPXC processes are running.**

IOW, because of various unimportant technical limitations:
- If you restart the client application, you have to re-authorize it.
- If you restart KeePassXC, you have to reauthorize each client application that wants access.

and, from the same comment:

> You can simply disable "Confirm when passwords are retrieved by clients" to revert to 2.6.x behavior for now.

That's all there is to it.

It's a known issue, tracked in this ticket, and not trivial to fix.

**Comment by KitchM** [comment]
Thanks everyone for the clarification.  Much appreciated.

If I understand, one brings up KPXC and goes to Settings>Secret Service Integration>"Confirm when passwords are retrieved by clients" and uncheck that box.  Then the environment will not display any confirmation box to allow an application to use the service.  Is that correct?

**Comment by michaelk83** [comment]
That's the idea, yes. Though there's another comment here suggesting this setting may be broken as well. (Which I forgot about earlier.) But you can go ahead and try it.

There is some security risk associated with disabling confirmation. So it's a choice between security and convenience.

**Comment by KitchM** [comment]
Thank you.

The way I see it, if I am using Linux not as root, there is little risk of outside control or compromise.  However, a firewall is handy to take care of further intrusion risks.  There may be other inherent factors of protection as well.

No one wants to know of every single thing that might be occurring unless it is important and/or constitutes a risk.  It is also unlikely that the average user will also be a security expert.  And even if that were the case, no body is perfect and mistakes may happen.

So where does that leave the user?  They must rely upon the open-source software to protect them.  If there is a further piece of software that will add another layer of protection, all the better; just install and move on.

This problem of continual concern over misleading information is a total waste of time.

With that said, non-standard acronyms such as FDO are a menace to people.  When we discuss these otherwise simple issues, we must be clear.  Further, when software processes change from the expected, there is an issue with programming.  For instance, when such actions were not required before, I hope that people understand that requiring it after seems downright weird.  Making these continual pop-ups as a default cannot be justified, IMHO.

Let's just hope the solutions given actually work over time.

Again, thanks everyone.

**Comment by siddhpant** [comment]
Edit: Sorry everyone, my bad, I made the mistake of unhelpfully ranting about an obvious GitHub troll in retrospect and doing the exact same thing I ranted about :( Kindly excuse that as a lack of experience.

---

@KitchM Please stop elongating for the sake of it with unrelated "gobbledegoop" as you say. I open my emails seeing the title in the notification in the hope something has actually progressed, and it has been mildly annoying to see this for past few replies.

Let me make everything clear for you.

**Using keepassxc as the secret service is a completely opt-in feature.** An average user might not know about it. It doesn't affect you if you don't know about it. Keepassxc doesn't run secret service by default. Keepassxc doesn't advertise on your face secret service integration by default. It is buried under settings.

A user on his own has to go and enable the secret service integration on his machine, which is a non-trivial process on Linux because your desktop environment probably already has a secret service running (like GNOME).

The secret service is a service used to fetch from keyring.

If you just want to have keyring access without confirmation, and don't want to bother with keepassxc, that's already happening. Nothing has changed by installing keepassxc.

Most people who want to use keepassxc secret service is because of the opt-in confirmation feature of the opt-in secret service integration, which is not the default on DEs like GNOME. GNOME allows all processes to access the keyring upon login. GNOME's security policy is of the session, so they assume the session is secure. Those who use keepassxc with the confirmation dialogue want a stricter control over access to keyring.

**This issue is solely for people using the confirmation dialogue.** Please stop derailing the discussion.

**Comment by KitchM** [comment]
@siddhpant So sorry you feel that way.  And how much did your rant help?  It might help you to turn off your email, since it is not required.  ;)
